### PR TITLE
Ensure override order of environment and other runtime settings

### PR DESCRIPTION
**Pull request by TomekTrzeciak**
Cylc uses `OrderedDict` to store configuration items in their definition order. However, when combined with family inheritance the resulting order can be surprising (from the the user perspective). Consider the following example:

```
[scheduling]
   [[dependencies]]
       graph = task
[runtime]
    [[FAM]]
        [[[environment]]]
            BAR = bar
    [[task]]
        inherit = FAM
        [[[environment]]]
            FOO = foo
            BAR = ${FOO}
```

This will store environment of `task` as:

```
$ cylc get-config --sparse -i '[runtime][task][environment]' SUITENAME
BAR = ${FOO}
FOO = foo
```

Note that `BAR` variable is now before `FOO`, unlike in the user definition, which will lead to unbound variable error in the job script. The reason for this is the definition of `BAR` under `FAM`. Even though `BAR` is subsequently redefined under `task`, the order of variables is determined by their first assignment rather than the last.

To look at it the other way, if `BAR` is not defined under `FAM`, the suite works OK. Then at a later time you add `BAR` under `FAM` (because it might be needed by some other task(s) inheriting from `FAM`) and the suite suddenly breaks and it's not very intuitive why.

I put this PR for discussion if this should be changed, as I've come across this problem more than once already.

This is a small change with no associated Issue.

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [ ] No change log entry required (why? e.g. invisible to users).
- [x] (7.8.x branch) I have updated the documentation in this PR branch (bug fix, no changes to docs needed).
- [ ] Add an equivelent PR for Master if this is approved.

### CODE REVIEW

**Comment by wxtim** [comment]
I've added myself as a reviewer because this relates to work I am doing for cylc8.

**Comment by TomekTrzeciak** [comment]
> I've added myself as a reviewer because this relates to work I am doing for cylc8.

Excellent! But note that this PR is against 7.8.x branch (although it should be applicable to master too, but not without some reshuffling).

File: tests/inheritance/03-override-order.t [hunk 1]
```diff
@@ -0,0 +1,44 @@
+#!/bin/bash
```
Comment (wxtim) [inline tests/inheritance/03-override-order.t]: My only concern is whether we might prefer to start a unit test of `replicate` using pytest. But this works.

Comment (TomekTrzeciak) [reply tests/inheritance/03-override-order.t]: Fair point, unit tests are cheaper.

**Comment by TomekTrzeciak** [comment]
Actually, based on [documentation](https://cylc.github.io/doc/built-sphinx/appendices/suiterc-config-ref.html#cylc-environment-variable):

> Values may refer to other local environment variables (order of definition is preserved)

I think the current behaviour should be treated as a bug.

**Comment by hjoliver** [comment]
@TomekTrzeciak -

Currently *the **original** definition order is preserved* - so if you override an inherited setting, you override it in its original place in the order.

To my mind, "original definition order" is what is usually meant by "definition order".  In your example `BAR` is defined in the parent, then overridden (not defined) in the child.

I think your way probably will cause less confusion, but I'm not sure it's a slam dunk.  Is the following example too contrived?:
```
[runtime]
    [[GREETER]]
        # Greet members by name.
        script = "echo $GREETING"
        [[[environment]]]
            NAME = "<OVERRIDE-ME>"
            GREETING = "HELLO ${NAME}"
    [[task]]
        inherit = GREETER
        [[[environment]]]
            NAME = bob
```
This works fine on master, but is broken on your branch!  

@cylc/core - any strong opinions on which way we should go here?

**Comment by oliver-sanders** [comment]
> any strong opinions on which way we should go here?

Of the two options I think original order makes more sense, the only workarounds I can think of the suit both cases are:

1. Use regex to pick out variable names and order accordingly
2. Don't override environment variables, dump both sections in namespace order i.e:

    ```bash
    # FAM
    BAR='bar'
    # task
    FOO='foo'
    BAR="${FOO}"
    ```

   But this would be kinda confusing in the job file (say for trigger edit) and could lead to problems where subshells are put in environment variables.

**Comment by dpmatthews** [comment]
I also think original order makes sense. I'd also be concerned about changing the behaviour and breaking existing workflows.

**Comment by TomekTrzeciak** [comment]
@hjoliver 
> Currently _the **original** definition order is preserved_ - so if you override an inherited setting, you override it in its original place in the order.

What is the definition order in your example if you flip definition of family and task around? The suite will work all the same, regardless of the order of runtime items. I think strictly speaking it is the inheritance order that is preserved and takes precedence over definition order (which I take literally as the order of items in the suite config file).

> To my mind, "original definition order" is what is usually meant by "definition order".

At the very least, more precise documentation is needed. Your definition of "definition order" is not the first one that springs to my mind. I find it rather quite counter-intuitive that the order specified 'locally' in the task's section (in front of my eyes, so to speak) is affected by inheritance from 'remote' definition in some family. There is increased cognitive pressure to reason about it in this way.

@oliver-sanders 
> 1. Use regex to pick out variable names and order accordingly

Probably impossible without reimplementing (part of) bash parser (bash variable expansions can be quite complicated). And than you need to consider the dependency order, which might be circular in general and thus not decidable. Lots of corner cases possible. Not worth the pain IMO.

> 2. Don't override environment variables, dump both sections in namespace order i.e:

Interesting. This would allow to incrementally prepend or append variables like PATH (c.f. #1363).

```
[runtime]
    [[PKG1]]
        [[[environment]]]
            PKG1_ROOT = /somewhere/here
            PATH = $PKG1_ROOT/bin:$PATH
    [[PKG2]]
        [[[environment]]]
            PKG2_ROOT = /over/there
            PATH = $PKG2_ROOT/bin:$PATH
    [[foo]]
        inherit = PKG1, PKG2
        script = pkg1_prog && pkg2_prog
```

>    But this would be kinda confusing in the job file (say for trigger edit) 

Provided that each inherited family is clearly delineated in the job script as in your example, I think this wouldn't be too confusing. The only downside I see is that this might require treating environment sections in a special way while parsing the config (or maybe it doesn't?). Plus this could be a potentially breaking change (as is this PR).

> and could lead to problems where subshells are put in environment variables.

Is that any different to the current situation (there's nothing stopping you from putting subshells in environment variables ATM).

**Comment by oliver-sanders** [comment]
> and could lead to problems where subshells are put in environment variables.

>> Is that any different to the current situation (there's nothing stopping you from putting subshells in environment variables ATM).

There could be multiple subshell calls per task e.g:

```ini
[runtime]
    [[FAM]]
        [[[environment]]]
             FOO = $(mktemp)
    [[task]]
        inherit = FAM
        [[[environment]]]
             FOO = $(mktemp -d)
```

**Comment by TomekTrzeciak** [comment]
> There could be multiple subshell calls per task e.g:

Oh, now I see what you mean. There might be undesired side effects if multiple environment sections are executed. Fair point.

OK, feel free to close this PR if you think the risk of breakage is too high for a patch release. I would still hope this could be addressed in 8.0 major release, though.

**Comment by matthewrmshin** [comment]
Looks like the consent is to leave the current behaviour as-is.

**Comment by TomekTrzeciak** [comment]
> Looks like the consent is to leave the current behaviour as-is.

@cylc/core, any opinions on addressing it in 8.0?

**Comment by oliver-sanders** [comment]
Could raise an issue.

**Comment by TomekTrzeciak** [comment]
> Could raise an issue.

#3332

**Pull request closed without merging**